### PR TITLE
Fix for the "next_deadline" bug.

### PR DESCRIPTION
**Pull request by serges147**
- 1st commit is just unit test which reproduces the problem
- 2nd commit - the fix, so CI is green again.

### CODE REVIEW

File: tests/test_embedded_scheduler.cpp [hunk 1]
```diff
@@ -323,6 +333,54 @@ TEST(TestEmbeddedScheduler, EventLoopPoll)
...
+    EventLoop<SteadyClockMock> evl;
+
+    auto evt = evl.defer(SteadyClockMock::now() + 1000ms, [&](auto) {});
+    EXPECT_TRUE(evt);
+
+    // This is special case - only one deferred event (and no "repeat"-s!), and it is already overdue (by +30ms).
+    // So, `next_deadline` should be `time_point::max()` b/c there will be nothing left pending after spin.
+    SteadyClockMock::advance(1000ms + 30ms);
+    const auto out = evl.spin();
+    EXPECT_THAT(out.next_deadline.time_since_epoch(), time_point::max().time_since_epoch());
```
Comment (serges147) [inline tests/test_embedded_scheduler.cpp:350]: At this line @ 350 we've got incorrect next deadline time before the fix (1000ms instead of correct max).

File: tests/test_embedded_scheduler.cpp [hunk 2]
```diff
@@ -323,6 +333,54 @@ TEST(TestEmbeddedScheduler, EventLoopPoll)
...
+        // so it will be already overdue for the next "b" event - should be executed as well.
+        calls.emplace_back("a", arg.deadline.time_since_epoch(), arg.approx_now.time_since_epoch());
+        SteadyClockMock::advance(100ms);
+    });
+    auto evt_b = evl.defer(SteadyClockMock::now() + 20ms, [&](const auto& arg) {  //
+        calls.emplace_back("b", arg.deadline.time_since_epoch(), arg.approx_now.time_since_epoch());
+    });
+
+    const auto out = evl.spin();
+    EXPECT_THAT(out.next_deadline.time_since_epoch(), duration::max());
```
Comment (serges147) [inline tests/test_embedded_scheduler.cpp:375]: And @ 375 as well.

File: include/embedded_scheduler/scheduler.hpp [hunk 3]
```diff
@@ -301,32 +301,41 @@ class EventLoop final
...
     {
-        auto next  = time_point::max();
-        auto now   = Clock::now();
-        auto worst = duration::zero();
+        SpinResult<Clock> result{.next_deadline  = time_point::max(),
+                                 .worst_lateness = duration::zero(),
+                                 .approx_now     = time_point::min()};
+        if (tree_.empty())
+        {
+            result.approx_now = Clock::now();
```
Comment (serges147) [inline include/embedded_scheduler/scheduler.hpp:309]: In original implementation, we always had 2 consecutive `now()` calls on the very first loop iteration for the future  scheduled event. Now we do only minimum (1) call per iteration.

File: include/embedded_scheduler/scheduler.hpp [hunk 4]
```diff
@@ -301,32 +301,41 @@ class EventLoop final
...
-            if (lateness < duration::zero())    // Too early -- either we need to sleep or the time sample is obsolete.
+            // The deadline is guaranteed to be set because it is in the tree.
+            const auto deadline = evt->getDeadline().value();
+            if (result.approx_now < deadline)  // Too early -- either we need to sleep or the time sample is obsolete.
             {
-                now      = Clock::now();  // The number of calls to Clock::now() is minimized.
-                lateness = now - next;
-                if (lateness < duration::zero())  // Nope, even with the latest time sample we are still early -- exit.
+                result.approx_now = Clock::now();  // The number of calls to Clock::now() is minimized.
+                if (result.approx_now < deadline)  // Nope, even with the latest time sample we are still early -- exit.
```
Comment (serges147) [inline include/embedded_scheduler/scheduler.hpp:320]: Note, that I also eliminated subtraction (which is over/under-flow prone strictly speaking). Now I subtract later, when I know for sure that `now >= deadline` (at line @ 333). Just in case, I will add separate unit test with deadline scheduling time near to `timepoint::max` - probably it will overflow even now...

File: tests/test_embedded_scheduler.cpp [hunk 5]
```diff
@@ -214,12 +222,13 @@ TEST(TestEmbeddedScheduler, EventLoopBasic)
...
-    EXPECT_THAT(50ms, out.worst_lateness);
-    EXPECT_THAT(14'000ms, evl->getTree()[0U]->getDeadline().value().time_since_epoch());
+    EXPECT_THAT(out.next_deadline.time_since_epoch(), 14'000ms);  // B removed so the next one is A.
+    EXPECT_THAT(out.worst_lateness, 50ms);
+    EXPECT_THAT(out.approx_now.time_since_epoch(), 13'050ms);
+    EXPECT_THAT(evl->getTree()[0U]->getDeadline().value().time_since_epoch(), 14'000ms);
     EXPECT_THAT(1, evl->getTree().size());  // Second dropped.
     EXPECT_TRUE(a);
-    EXPECT_THAT(13'000ms, a.value().deadline.time_since_epoch());
+    EXPECT_THAT(a.value().deadline.time_since_epoch(), 13'000ms);
```
Comment (serges147) [inline tests/test_embedded_scheduler.cpp:231]: In gtest it is always **first actual** value, and 2nd-ly the **expected** - result output (in case of failure) explicitly mentions what was expected and was actually found).

File: include/embedded_scheduler/scheduler.hpp (hunk 3 above)
Comment (pavel-kirienko) [reply include/embedded_scheduler/scheduler.hpp:309]: Please tag this branch with `[[unlikely]]`. Maybe there are other places that could use this.

**Review by pavel-kirienko — APPROVED** [review]
